src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesClientProvider.java

```diff
@@ -0,0 +1,79 @@
+package org.csanchez.jenkins.plugins.kubernetes;
+
+import com.google.common.base.Objects;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.RemovalListener;
+
+import io.fabric8.kubernetes.client.KubernetesClient;
+
+import javax.annotation.CheckForNull;
+import java.io.IOException;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateEncodingException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Manages the Kubernetes client creation per cloud
+ */
+final class KubernetesClientProvider {
+
+    private static final Cache<String, Client> clients;
+
+    private static final Integer CACHE_SIZE;
+    private static final Integer CACHE_TTL;
+
+    static {
+        CACHE_SIZE = Integer.getInteger("org.csanchez.jenkins.plugins.kubernetes.clients.cacheSize", 10);
+        CACHE_TTL = Integer.getInteger("org.csanchez.jenkins.plugins.kubernetes.clients.cacheTtl", 60);
+        clients = CacheBuilder.newBuilder()
+                .maximumSize(CACHE_SIZE)
+                .expireAfterWrite(CACHE_TTL, TimeUnit.MINUTES)
+                .removalListener((RemovalListener<String, Client>) removalNotification -> {
+                    removalNotification.getValue().getClient().close();
+                })
+                .build();
+    }
+
+    private KubernetesClientProvider() {
+    }
+
+    static KubernetesClient createClient(final String cloudName, final String serviceAddress, final String namespace, @CheckForNull final String caCertData,
+                                                @CheckForNull final String credentials, final boolean skipTlsVerify, final int connectTimeout, final int readTimeout, final int maxRequestsPerHost) throws NoSuchAlgorithmException, UnrecoverableKeyException,
+            KeyStoreException, IOException, CertificateEncodingException, ExecutionException {
+
+        final int validity = Objects.hashCode(serviceAddress, namespace, caCertData, credentials, skipTlsVerify, connectTimeout, readTimeout, maxRequestsPerHost);
+        final Client c = clients.getIfPresent(cloudName);
+        if (c != null && validity == c.getValidity()) {
+            return c.getClient();
+        } else {
```

Comment:
now we are closing connections in the listener, probably we should close also if c!=null && validity != c.getValidity() before creating a new KubernetesClient

Reply:
it would be good to close existing client, because `onRemoval` could be called later in time

https://google.github.io/guava/releases/23.0/api/docs/com/google/common/cache/RemovalListener.html#onRemoval-com.google.common.cache.RemovalNotification-

> Notifies the listener that a removal occurred at some point in the past.